### DISCUSSION TITLE
Updating MathTex using a ValueTracker

### DISCUSSION THREAD

**Discussion by VladislavKorecky** [ManimCommunity/manim#3327] [Q&A] [answered]
I want to update a `MathTex` mobject, using a `ValueTracker`. I came up with the following code, but the text isn't updating:

```python
from manim import *


class Parameters(Scene):
    def __init__(self):
        super().__init__()

    def construct(self):
        a = ValueTracker(1)
        b = ValueTracker(0)

        a_targets = [4, -2, 1]
        b_targets = [1, -1, 0]

        a_text = MathTex(f"a = {a.get_value()}").move_to(UP * 0.5)
        b_text = MathTex(f"b = {b.get_value()}").move_to(DOWN * 0.5)

        a_text.add_updater(
            lambda mob: mob.set_value(f"a = {a.get_value()}")
        )
        b_text.add_updater(
            lambda mob: mob.set_value(f"b = {b.get_value()}")
        )

        self.play(FadeIn(a_text), FadeIn(b_text))
        self.wait(0.5)

        for target in a_targets:
            self.play(a_text.animate.set_value(target), rate_func=rate_functions.linear)
            self.wait(0.5)

        for target in b_targets:
            self.play(b_text.animate.set_value(target), rate_func=rate_functions.linear)
            self.wait(0.5)

```

I ran the code with the following command: `manim -p -ql --disable_caching parameters.py`

**Comment by uwezi** [comment, marked as answer]
You cannot just set the "value" of a `MathTex` object and expect something to happen. `MathTex()` and related objects are sent to LaTeX as the external program to do the actual conversion only when they are first created. Afterward you cannot really change how they look. Also I am not quite sure if you are really after the final looks with a varying number of decimal places... you could add format specifiers to your python code though...

So in your case you can use your updaters together with the `.become` method, or use `always_redraw` from the beginning. And that's apart from the fact that you are _animating_ the wrong objects in your `self.play`.
And then there is a much more suitable object for this purpose, the `Variable` - I added it as an example to your code.

For a better help forum come over to [Discord](https://docs.manim.community/en/stable/faq/general.html?highlight=discord#where-can-i-find-more-resources-for-learning-manim)

```python
class Parameters(Scene):
    # I don't think this is a good idea!
    #def __init__(self):
    #    super().__init__()

    def construct(self):
        a = ValueTracker(1)
        b = ValueTracker(0)

        a_targets = [4, -2, 1]
        b_targets = [1, -1, 0]

        a_text = MathTex(f"a = {a.get_value()}").move_to(UP * 0.5)

        b_text = always_redraw(lambda: 
            MathTex(f"b = {b.get_value()}").move_to(DOWN * 0.5)
        )

        a_text.add_updater(
            lambda mob: mob.become(MathTex(f"a = {a.get_value()}").move_to(UP * 0.5))
        )

        c_everything = Variable(0, label="c", num_decimal_places=2).move_to(DOWN * 1.5)

        self.play(FadeIn(a_text), FadeIn(b_text), FadeIn(c_everything))
        self.wait(0.5)

        for target in a_targets:
            self.play(a.animate.set_value(target), rate_func=rate_functions.linear)
            self.wait(0.5)

        for target in b_targets:
            self.play(b.animate.set_value(target), rate_func=rate_functions.linear)
            self.wait(0.5)        

        for target in a_targets:
            self.play(c_everything.tracker.animate.set_value(target), rate_func=rate_functions.linear)
            self.wait(0.5)                    
```

https://github.com/ManimCommunity/manim/assets/8582807/63e245a5-80aa-4e48-9080-4f1ccedb7d6a

**Reply by VladislavKorecky** [reply]
`Variable` was a perfect solution. Didn't even know something like this existed. Thanks a lot.

**Reply by VladislavKorecky** [reply]
Also, how can I change the font size of the `Variable`? Both for the label and the value.

**Reply by uwezi** [reply]
First you can always just scale the whole object to whichever size you want - this is much more intuitive than selecting arbitrary font sizes:  ```Variable(0, label="c", num_decimal_places=2).scale_to_fit_height(0.5).move_to(DOWN * 1.5)``` scales the object to 0.5 height units = 1/16 of the screen height.

https://github.com/ManimCommunity/manim/assets/8582807/7a0e2ac5-68cb-4f66-a1b9-b45f54a10dc0

You could also create new object classes for both the label and the number, where you could modify the font size, interestingly the `Variable()` object itself does not accept a `font_size=` argument, nor a dictionary over sub-mobject properties like other compound object (`NumberLine` for example).

```python
class VariableExample(Scene):
    def construct(self):
        start = 2.0

        x_var = Variable(start, 'x', num_decimal_places=3).scale_to_fit_height(1)
        sqr_var = Variable(start**2, 'x^2', num_decimal_places=3).scale_to_fit_height(0.5)
        Group(x_var, sqr_var).arrange(DOWN)

        sqr_var.add_updater(lambda v: v.tracker.set_value(x_var.tracker.get_value()**2))

        self.add(x_var, sqr_var)
        self.play(x_var.tracker.animate.set_value(5), run_time=2, rate_func=linear)
        self.wait(0.1)  
```

**Reply by VladislavKorecky** [reply]
Got it. Thanks mate.
